src/main/java/com/redhat/rhjmc/containerjfr/net/TargetConnectionManager.java

```diff
@@ -63,42 +67,72 @@
     public static final Pattern HOST_PORT_PAIR_PATTERN =
             Pattern.compile("^([^:\\s]+)(?::(\\d{1,5}))?$");
 
-    private final Logger logger;
-    // FIXME verify concurrent connection safety and remove locking
-    private final ReentrantLock lock = new ReentrantLock();
-    // maintain a short-lived cache of connections to allow nested ConnectedTasks
-    // without having to manage connection reuse
-    private final Map<ConnectionDescriptor, JFRConnection> activeConnections = new HashMap<>();
+    static final Duration DEFAULT_TTL = Duration.ofSeconds(90);
+
     private final Lazy<JFRConnectionToolkit> jfrConnectionToolkit;
+    private final Logger logger;
 
-    public TargetConnectionManager(Logger logger, Lazy<JFRConnectionToolkit> jfrConnectionToolkit) {
-        this.logger = logger;
+    private final LoadingCache<ConnectionDescriptor, JFRConnection> connections;
+
+    TargetConnectionManager(
+            Lazy<JFRConnectionToolkit> jfrConnectionToolkit, Duration ttl, Logger logger) {
         this.jfrConnectionToolkit = jfrConnectionToolkit;
+        this.logger = logger;
+
+        this.connections =
+                Caffeine.newBuilder()
+                        .scheduler(Scheduler.systemScheduler())
+                        .expireAfterAccess(ttl)
+                        .removalListener(
+                                new RemovalListener<ConnectionDescriptor, JFRConnection>() {
+                                    @Override
+                                    public void onRemoval(
+                                            ConnectionDescriptor descriptor,
+                                            JFRConnection connection,
+                                            RemovalCause cause) {
+                                        if (descriptor == null) {
+                                            logger.warn(
+                                                    "Connection eviction triggered with null descriptor");
+                                            return;
+                                        }
+                                        if (connection == null) {
+                                            logger.warn(
+                                                    "Connection eviction triggered with null connection");
+                                            return;
+                                        }
+                                        logger.info(
+                                                String.format(
```

Comment:
This can be written as logger.info("removing cached connection for %s", descriptor.getTargetId()). I changed all the string.format statements in a PR a month ago, it probably makes sense to keep it consistent!

Reply:
Aha, good catch - this PR was opened as a draft the day before you opened that logger PR, so I just never came back around to fix that up :-)